src/fsharp/FSharp.Core/seqcore.fs

```diff
@@ -400,3 +404,110 @@ namespace Microsoft.FSharp.Core.CompilerServices
             member x.MoveNext() = x.MoveNextImpl()
 
             member _.Reset() = raise <| new System.NotSupportedException()
+
+
+    [<Struct; NoEquality; NoComparison>]
+    type ListCollector<'T> =
+        [<DefaultValue(false)>]
+        val mutable Result : 'T list
+
+        [<DefaultValue(false)>]
+        val mutable LastCons : 'T list
+
+        member this.Add (value: 'T) =
+            match box this.Result with 
+            | null -> 
+                let ra = RuntimeHelpers.FreshConsNoTail value
+                this.Result <- ra
+                this.LastCons <- ra
+            | _ -> 
+                let ra = RuntimeHelpers.FreshConsNoTail value
+                RuntimeHelpers.SetFreshConsTail this.LastCons ra
+                this.LastCons <- ra
+
+        member this.AddMany (values: seq<'T>) =
+            // cook a faster iterator for lists and arrays
+            match values with 
+            | :? ('T[]) as valuesAsArray -> 
+                for v in valuesAsArray do
+                   this.Add v
```

Comment:
It's possible this could be a bit faster and avoid so many writes into the fields of ListCollector. However ListCollector is ultimately a mutable struct on the stack, so writes will be fast, it may not end up any faster